lib/datadog/grok/src/matchers/date.rs

```diff
@@ -0,0 +1,210 @@
+    let mut time_format = String::new();
+    let mut chars = format.chars().peekable();
+    while let Some(&c) = chars.peek() {
+        if ('A'..='Z').contains(&c) || ('a'..='z').contains(&c) {
+            let token: String = chars.by_ref().peeking_take_while(|&cn| cn == c).collect();
+            match token.chars().next().unwrap() {
+                // hour of day (number, 1..12)
+                'h' => time_format.push_str("%I"),
+                // hour of day (number, 0..23)
+                'H' => time_format.push_str("%H"),
+                //  minute of hour
+                'm' => time_format.push_str("%M"),
+                // second of minute
+                's' => time_format.push_str("%S"),
+                // fraction of second
+                'S' => time_format.push_str("%3f"),
+                // year
+                'y' | 'Y' => time_format.push_str("%Y"),
+                // weekyear
+                'x' => time_format.push_str("%D"),
+                // century
+                'c' | 'C' => time_format.push_str("%C"),
+                // day
+                'd' => time_format.push_str("%d"),
+                // day of week
+                'e' => time_format.push_str("%u"),
+                // day of year
+                'D' => time_format.push_str("%j"),
+                // week of year
+                'w' => time_format.push_str("%V"),
+                // month of year
+                'M' => {
+                    if token.len() == 2 {
+                        // Month number
+                        time_format.push_str("%m");
+                    } else if token.len() == 3 {
+                        // Abbreviated month name. Always 3 letters.
+                        time_format.push_str("%b");
+                    } else if token.len() > 3 {
+                        // Full month name
+                        time_format.push_str("%B");
+                    }
+                }
+                // AM/PM
+                'a' => time_format.push_str("%p"),
+                // dayOfWeek (text)
+                'E' if token.len() == 3 => time_format.push_str("%a"),
+                'E' if token.len() > 3 => time_format.push_str("%A"),
+                // time zone (text)
+                'z' => time_format.push_str("%Z"),
+                // time zone offset
+                'Z' => {
+                    if token.len() == 1 {
+                        time_format.push_str("%z");
+                    } else if token.len() == 2 {
+                        time_format.push_str("%:z");
+                    }
+                }
+                _ => return Err(format!("invalid date format '{}'", format)),
+            }
+        } else if c == '\''
+        // quoted literal
+        {
+            let literal: String = chars
+                .by_ref()
+                .skip(1)
+                .take_while(|&cn| cn != '\'')
+                .collect();
```

Comment:
Does this need to take into account escapes?

So things like `..'thing\'thang'..`..

Reply:
No, it doesn't. It's only to support expressions like this: "yyyy-MM-dd'T'HH:mm:ss.SSSZ"